Objects/object.c

```diff
@@ -2107,6 +2107,18 @@ _PyTrash_thread_destroy_chain(void)
          * up distorting allocation statistics.
          */
         assert(op->ob_refcnt == 0);
+        /* We need to increase trash_delete_nesting here, otherwise,
+           _PyTrash_thread_destroy_chain will be called recursively
+           and then possibly crash.  An example that may crash without
+           increase:
+               N = 500000  # need to be large enough
+               ob = object()
+               tups = [(ob,) for i in range(N)]
+               for i in range(49):
+                   tups = [(tup,) for tup in tups]
+               del tups
+        */
+        assert(tstate->trash_delete_nesting == 0);
         ++tstate->trash_delete_nesting;
```

Comment:
It may be safe to move it out of the loop indeed. But I'm not sure it's a good idea to touch this code too much, either.